File: frontend/lib/circuit-breaker.ts
/**
 * Circuit Breaker Pattern Implementation
 *
 * Prevents cascading failures by stopping calls to failing services.
 * States: CLOSED (normal) -> OPEN (failing) -> HALF_OPEN (testing)
 *
 * @example
 * const breaker = new CircuitBreaker("anthropic", { failureThreshold: 5 });
 * const result = await breaker.call(() => anthropic.messages.create({...}));
 */

export type CircuitState = "CLOSED" | "OPEN" | "HALF_OPEN";

export interface CircuitBreakerOptions {
  /** Number of failures before opening circuit (default: 5) */
  failureThreshold?: number;
  /** Time in ms before attempting recovery (default: 60000 = 1 min) */
  recoveryTimeout?: number;
  /** Number of successful calls in HALF_OPEN to close circuit (default: 2) */
  successThreshold?: number;
  /** Timeout for individual calls in ms (default: 30000 = 30s) */
  callTimeout?: number;
  /** Called when circuit opens */
  onOpen?: (name: string, error: Error) => void;
  /** Called when circuit closes */
  onClose?: (name: string) => void;
  /** Called when circuit enters half-open */
  onHalfOpen?: (name: string) => void;
}

export interface CircuitBreakerStats {
  state: CircuitState;
  failures: number;
  successes: number;
  lastFailureTime: number | null;
  totalCalls: number;
  totalFailures: number;
  totalSuccesses: number;
}

export class CircuitBreakerError extends Error {
  constructor(
    message: string,
    public readonly circuitName: string,
    public readonly state: CircuitState
  ) {
    super(message);
    this.name = "CircuitBreakerError";
  }
}

export class CircuitBreaker {
  private state: CircuitState = "CLOSED";
  private failures = 0;
  private successes = 0;
  private lastFailureTime: number | null = null;
  private totalCalls = 0;
  private totalFailures = 0;
  private totalSuccesses = 0;

  private readonly failureThreshold: number;
  private readonly recoveryTimeout: number;
  private readonly successThreshold: number;
  private readonly callTimeout: number;
  private readonly onOpen?: (name: string, error: Error) => void;
  private readonly onClose?: (name: string) => void;
  private readonly onHalfOpen?: (name: string) => void;

  constructor(
    private readonly name: string,
    options: CircuitBreakerOptions = {}
  ) {
    this.failureThreshold = options.failureThreshold ?? 5;
    this.recoveryTimeout = options.recoveryTimeout ?? 60000;
    this.successThreshold = options.successThreshold ?? 2;
    this.callTimeout = options.callTimeout ?? 30000;
    this.onOpen = options.onOpen;
    this.onClose = options.onClose;
    this.onHalfOpen = options.onHalfOpen;
  }

  /**
   * Execute a function through the circuit breaker.
   * Throws CircuitBreakerError if circuit is open.
   */
  async call<T>(fn: () => Promise<T>): Promise<T> {
    this.totalCalls++;

    // Check if we should attempt recovery
    if (this.state === "OPEN") {
      if (this.shouldAttemptRecovery()) {
        this.transitionTo("HALF_OPEN");
      } else {
        throw new CircuitBreakerError(
          `Circuit breaker "${this.name}" is OPEN. Service temporarily unavailable.`,
          this.name,
          this.state
        );
      }
    }

    try {
      const result = await this.executeWithTimeout(fn);
      this.onSuccess();
      return result;
    } catch (error) {
      this.onFailure(error as Error);
      throw error;
    }
  }

  /**
   * Execute function with timeout protection.
   */
  private async executeWithTimeout<T>(fn: () => Promise<T>): Promise<T> {
    return new Promise<T>((resolve, reject) => {
      const timeoutId = setTimeout(() => {
        reject(new Error(`Circuit breaker "${this.name}" call timed out after ${this.callTimeout}ms`));
      }, this.callTimeout);

      fn()
        .then((result) => {
          clearTimeout(timeoutId);
          resolve(result);
        })
        .catch((error) => {
          clearTimeout(timeoutId);
          reject(error);
        });
    });
  }

  /**
   * Handle successful call.
   */
  private onSuccess(): void {
    this.totalSuccesses++;

    if (this.state === "HALF_OPEN") {
      this.successes++;
      if (this.successes >= this.successThreshold) {
        this.transitionTo("CLOSED");
      }
    } else if (this.state === "CLOSED") {
      // Reset failure count on success in closed state
      this.failures = 0;
    }
  }

  /**
   * Handle failed call.
   */
  private onFailure(error: Error): void {
    this.totalFailures++;
    this.failures++;
    this.lastFailureTime = Date.now();

    if (this.state === "HALF_OPEN") {
      // Any failure in half-open goes back to open
      this.transitionTo("OPEN", error);
    } else if (this.state === "CLOSED" && this.failures >= this.failureThreshold) {
      this.transitionTo("OPEN", error);
    }
  }

  /**
   * Check if enough time has passed to attempt recovery.
   */
  private shouldAttemptRecovery(): boolean {
    if (!this.lastFailureTime) return true;
    return Date.now() - this.lastFailureTime >= this.recoveryTimeout;
  }

  /**
   * Transition to a new state.
   */
  private transitionTo(newState: CircuitState, error?: Error): void {
    const previousState = this.state;
    this.state = newState;

    // Reset counters on state change
    if (newState === "CLOSED") {
      this.failures = 0;
      this.successes = 0;
      this.onClose?.(this.name);
      console.log(`[CircuitBreaker] ${this.name}: ${previousState} -> CLOSED (recovered)`);
    } else if (newState === "OPEN") {
      this.successes = 0;
      this.onOpen?.(this.name, error!);
      console.warn(`[CircuitBreaker] ${this.name}: ${previousState} -> OPEN (failures: ${this.failures})`);
    } else if (newState === "HALF_OPEN") {
      this.successes = 0;
      this.onHalfOpen?.(this.name);
      console.log(`[CircuitBreaker] ${this.name}: ${previousState} -> HALF_OPEN (attempting recovery)`);
    }
  }

  /**
   * Get current circuit breaker statistics.
   */
  getStats(): CircuitBreakerStats {
    return {
      state: this.state,
      failures: this.failures,
      successes: this.successes,
      lastFailureTime: this.lastFailureTime,
      totalCalls: this.totalCalls,
      totalFailures: this.totalFailures,
      totalSuccesses: this.totalSuccesses,
    };
  }

  /**
   * Manually reset the circuit breaker to CLOSED state.
   * Use with caution - typically for testing or manual intervention.
   */
  reset(): void {
    this.state = "CLOSED";
    this.failures = 0;
    this.successes = 0;
    this.lastFailureTime = null;
    console.log(`[CircuitBreaker] ${this.name}: Manually reset to CLOSED`);
  }

  /**
   * Check if circuit is currently allowing calls.
   */
  isAvailable(): boolean {
    if (this.state === "CLOSED") return true;
    if (this.state === "HALF_OPEN") return true;
    if (this.state === "OPEN" && this.shouldAttemptRecovery()) return true;
    return false;
  }

  /**
   * Get the current state.
   */
  getState(): CircuitState {
    return this.state;
  }
}

/**
 * Global registry for circuit breakers.
 * Allows sharing circuit breakers across modules.
 */
class CircuitBreakerRegistry {
  private breakers = new Map<string, CircuitBreaker>();

  /**
   * Get or create a circuit breaker by name.
   */
  get(name: string, options?: CircuitBreakerOptions): CircuitBreaker {
    let breaker = this.breakers.get(name);
    if (!breaker) {
      breaker = new CircuitBreaker(name, options);
      this.breakers.set(name, breaker);
    }
    return breaker;
  }

  /**
   * Get all circuit breaker stats.
   */
  getAllStats(): Record<string, CircuitBreakerStats> {
    const stats: Record<string, CircuitBreakerStats> = {};
    this.breakers.forEach((breaker, name) => {
      stats[name] = breaker.getStats();
    });
    return stats;
  }

  /**
   * Reset all circuit breakers.
   */
  resetAll(): void {
    this.breakers.forEach((breaker) => {
      breaker.reset();
    });
  }
}

export const circuitBreakerRegistry = new CircuitBreakerRegistry();

/**
 * Pre-configured circuit breakers for common services.
 */
export const circuitBreakers = {
  anthropic: circuitBreakerRegistry.get("anthropic", {
    failureThreshold: 3,
    recoveryTimeout: 60000,
    successThreshold: 2,
    callTimeout: 120000, // 2 minutes for AI calls
    onOpen: (name, error) => {
      console.error(`[AI Service] ${name} circuit opened:`, error.message);
    },
  }),

  stripe: circuitBreakerRegistry.get("stripe", {
    failureThreshold: 5,
    recoveryTimeout: 30000,
    successThreshold: 2,
    callTimeout: 30000,
    onOpen: (name, error) => {
      console.error(`[Billing] ${name} circuit opened:`, error.message);
    },
  }),

  backend: circuitBreakerRegistry.get("backend", {
    failureThreshold: 5,
    recoveryTimeout: 30000,
    successThreshold: 3,
    callTimeout: 60000,
    onOpen: (name, error) => {
      console.error(`[Backend API] ${name} circuit opened:`, error.message);
    },
  }),
};

File: backend/app/services/circuit_breaker.py
"""
Circuit breaker implementation for external API calls.

The circuit breaker pattern prevents cascading failures when external services
are unavailable or experiencing issues. It monitors failures and opens the
circuit to prevent further calls when a threshold is exceeded.

States:
- CLOSED: Normal operation, requests pass through
- OPEN: Failures exceeded threshold, requests fail fast
- HALF_OPEN: Testing if service recovered, limited requests allowed

Usage:
    breaker = CircuitBreaker("openai", failure_threshold=5, recovery_timeout=60)
    
    try:
        result = await breaker.call(async_api_function, *args, **kwargs)
    except CircuitBreakerOpen:
        # Handle circuit open state (service unavailable)
        pass
"""

import asyncio
import time
from dataclasses import dataclass, field
from enum import Enum
from functools import wraps
from typing import Any, Callable, Optional, TypeVar, ParamSpec
import structlog

logger = structlog.get_logger()

P = ParamSpec("P")
T = TypeVar("T")


class CircuitState(Enum):
    """Circuit breaker states."""
    CLOSED = "closed"      # Normal operation
    OPEN = "open"          # Failing fast
    HALF_OPEN = "half_open"  # Testing recovery


class CircuitBreakerOpen(Exception):
    """Raised when circuit breaker is open."""
    
    def __init__(self, service_name: str, retry_after: float):
        self.service_name = service_name
        self.retry_after = retry_after
        super().__init__(
            f"Circuit breaker open for {service_name}. Retry after {retry_after:.1f}s"
        )


@dataclass
class CircuitBreaker:
    """
    Circuit breaker for external service calls.
    
    Args:
        service_name: Name of the service (for logging)
        failure_threshold: Number of failures before opening circuit
        recovery_timeout: Seconds to wait before testing recovery
        success_threshold: Successes needed in half-open to close circuit
        expected_exceptions: Exception types to count as failures
    """
    service_name: str
    failure_threshold: int = 5
    recovery_timeout: float = 60.0
    success_threshold: int = 2
    expected_exceptions: tuple = field(default_factory=lambda: (Exception,))
    
    # State
    state: CircuitState = field(default=CircuitState.CLOSED, init=False)
    failure_count: int = field(default=0, init=False)
    success_count: int = field(default=0, init=False)
    last_failure_time: float = field(default=0.0, init=False)
    last_state_change: float = field(default_factory=time.time, init=False)
    
    def __post_init__(self):
        self._lock = asyncio.Lock()
    
    @property
    def is_closed(self) -> bool:
        return self.state == CircuitState.CLOSED
    
    @property
    def is_open(self) -> bool:
        return self.state == CircuitState.OPEN
    
    @property
    def is_half_open(self) -> bool:
        return self.state == CircuitState.HALF_OPEN
    
    @property
    def time_since_last_failure(self) -> float:
        return time.time() - self.last_failure_time
    
    @property
    def retry_after(self) -> float:
        """Seconds until circuit will transition to half-open."""
        if not self.is_open:
            return 0.0
        remaining = self.recovery_timeout - self.time_since_last_failure
        return max(0.0, remaining)
    
    async def _transition_to(self, new_state: CircuitState) -> None:
        """Transition to a new state."""
        old_state = self.state
        self.state = new_state
        self.last_state_change = time.time()
        
        if new_state == CircuitState.HALF_OPEN:
            self.success_count = 0
        elif new_state == CircuitState.CLOSED:
            self.failure_count = 0
            self.success_count = 0
        
        logger.info(
            "Circuit breaker state transition",
            service=self.service_name,
            old_state=old_state.value,
            new_state=new_state.value,
        )
    
    async def _record_success(self) -> None:
        """Record a successful call."""
        async with self._lock:
            if self.is_half_open:
                self.success_count += 1
                if self.success_count >= self.success_threshold:
                    await self._transition_to(CircuitState.CLOSED)
    
    async def _record_failure(self, error: Exception) -> None:
        """Record a failed call."""
        async with self._lock:
            self.failure_count += 1
            self.last_failure_time = time.time()
            
            logger.warning(
                "Circuit breaker recorded failure",
                service=self.service_name,
                failure_count=self.failure_count,
                threshold=self.failure_threshold,
                error=str(error)[:100],
            )
            
            if self.is_half_open:
                # Any failure in half-open immediately opens circuit
                await self._transition_to(CircuitState.OPEN)
            elif self.failure_count >= self.failure_threshold:
                await self._transition_to(CircuitState.OPEN)
    
    async def _check_state(self) -> None:
        """Check and potentially update circuit state."""
        async with self._lock:
            if self.is_open and self.time_since_last_failure >= self.recovery_timeout:
                await self._transition_to(CircuitState.HALF_OPEN)
    
    async def call(
        self,
        func: Callable[P, T],
        *args: P.args,
        **kwargs: P.kwargs,
    ) -> T:
        """
        Execute a function through the circuit breaker.
        
        Args:
            func: Async or sync function to call
            *args: Arguments to pass to function
            **kwargs: Keyword arguments to pass to function
            
        Returns:
            Result of the function call
            
        Raises:
            CircuitBreakerOpen: If circuit is open
            Exception: Any exception from the wrapped function
        """
        await self._check_state()
        
        if self.is_open:
            raise CircuitBreakerOpen(self.service_name, self.retry_after)
        
        try:
            # Handle both async and sync functions
            if asyncio.iscoroutinefunction(func):
                result = await func(*args, **kwargs)
            else:
                result = func(*args, **kwargs)
            
            await self._record_success()
            return result
            
        except self.expected_exceptions as e:
            await self._record_failure(e)
            raise
    
    def get_status(self) -> dict:
        """Get current circuit breaker status."""
        return {
            "service": self.service_name,
            "state": self.state.value,
            "failure_count": self.failure_count,
            "failure_threshold": self.failure_threshold,
            "success_count": self.success_count,
            "retry_after": self.retry_after,
            "time_since_last_failure": self.time_since_last_failure if self.last_failure_time else None,
        }
    
    def reset(self) -> None:
        """Manually reset the circuit breaker to closed state."""
        self.state = CircuitState.CLOSED
        self.failure_count = 0
        self.success_count = 0
        self.last_failure_time = 0.0
        self.last_state_change = time.time()
        
        logger.info("Circuit breaker manually reset", service=self.service_name)


# Global circuit breakers for external services
_circuit_breakers: dict[str, CircuitBreaker] = {}


def get_circuit_breaker(
    service_name: str,
    failure_threshold: int = 5,
    recovery_timeout: float = 60.0,
) -> CircuitBreaker:
    """
    Get or create a circuit breaker for a service.
    
    Uses a singleton pattern to ensure one breaker per service.
    """
    if service_name not in _circuit_breakers:
        _circuit_breakers[service_name] = CircuitBreaker(
            service_name=service_name,
            failure_threshold=failure_threshold,
            recovery_timeout=recovery_timeout,
        )
    return _circuit_breakers[service_name]


def with_circuit_breaker(
    service_name: str,
    failure_threshold: int = 5,
    recovery_timeout: float = 60.0,
):
    """
    Decorator to wrap a function with circuit breaker protection.
    
    Usage:
        @with_circuit_breaker("openai")
        async def call_openai_api(...):
            ...
    """
    def decorator(func: Callable[P, T]) -> Callable[P, T]:
        breaker = get_circuit_breaker(service_name, failure_threshold, recovery_timeout)
        
        @wraps(func)
        async def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
            return await breaker.call(func, *args, **kwargs)
        
        return wrapper
    return decorator


def get_all_circuit_breaker_statuses() -> list[dict]:
    """Get status of all circuit breakers for monitoring."""
    return [breaker.get_status() for breaker in _circuit_breakers.values()]
